minimal zsh completion

trws opened the pull request:
Since zsh can load bash completion files natively, seems reasonable to just turn this on.  The only changes are to switch from `type -t` which zsh doesn't support to using `type` with a regex and adding a new arm to the sourcing of the completions to allow it to work for zsh as well as bash.

Could use more bash/dash/etc testing probably, but everything I've thought to try has worked so far.

trws committed:
minimal zsh completion

trws committed:
update input file to match generated

tgamblin reviewed (changes requested):
not sure what, but something is breaking with the shell tests.

trws committed:
unit-test zsh support, fix issues

Specifically fixed word splitting in completion-test, use a different
method to apply sh emulation to zsh loaded bash completion, and fixed
an incompatibility in regex operator quoting requirements.

trws commented:
It turned out to be a difference in the requirements for quoting of the regex match operator.  In fixing it, I added zsh completion to the unit tests and fixed some follow-on bugs caused by needing bash-isms in the test harness and some functions that weren't getting used for interactive completion.

trws committed:
compinit now ignores insecure directories

Completion isn't meant to be enabled in non-interactive environments, so
by default compinit will ask the user if they want to ignore insecure
directories or load them anyway.  To pass the spack unit tests in GH
actions, this prompt must be disabled, so ignore explicitly until a
better solution can be found.

trws committed:
debug functions test also requires bash emulation

COMP_WORDS is a bash-ism that zsh doesn't natively support, turn on
emulation for just that section of tests to allow the comparison to
work.  Does not change the behavior of the functions themselves since
they are already pinned to sh emulation elsewhere.

trws committed:
propagate change to .in file

trws committed:
fix comment and update script based on .in

trws commented:
Ok, this is now passing all but coverage testing for the completion script.  As of now, kcov does not [support coverage for zsh](https://github.com/shellspec/shellspec/pull/64) except through a non-upstreamed [extension](https://github.com/shellspec/shellspec/pull/64).  The code is covered by unit testing, but not in a way that I can report to codecov at the moment.  Is that a blocking issue such that I should include the workaround to get zsh coverage in here as well?

trws commented:
@adamjstewart, @tgamblin, thoughts on the coverage issue?

adamjstewart reviewed (approved):
LGTM. It's too bad about the lack of coverage, but hopefully that will be resolved someday. Tests without coverage are better than no tests in my mind.

trws commented:
If you want, we can set up another issue to track this, and we could add it with the workaround above, it just isn't going to be officially supported by kcov.
